src/association/association_internal.rs

```
@@ -2175,7 +2192,68 @@ impl AssociationInternal {
+            match handle_code {
+                0b00 => {
+                    // Stop processing this packet
+                    return Err(Error::ErrChunkTypeUnhandled)
+                }
+                0b01 => {
+                    // stop processing but report the chunk as unrecognized
+                    let err_chunk = ChunkError {
+                        error_causes: vec![ErrorCause {
+                            code: UNRECOGNIZED_CHUNK_TYPE,
+                            raw: chunk.marshal()?,
+                        }]
+                    };
+                    let packet = Packet {
+                        verification_tag: self.peer_verification_tag,
+                        source_port: self.source_port,
+                        destination_port: self.destination_port,
+                        chunks: vec![Box::new(err_chunk)]
+                    };
+                    self.control_queue.push_back(packet);
+                    self.awake_write_loop();
+                    return Err(Error::ErrChunkTypeUnhandled)
+                }
+                0b10 => {
+                    // just ignore
+                    vec![]
+                }
+                0b11 => {
+                    // keep processing but report the chunk as unrecognized
+                    let err_chunk = ChunkError {
+                        error_causes: vec![ErrorCause {
+                            code: UNRECOGNIZED_CHUNK_TYPE,
+                            raw: chunk.marshal()?,
+                        }]
+                    };
+                    let packet = Packet {
+                        verification_tag: self.peer_verification_tag,
+                        source_port: self.source_port,
+                        destination_port: self.destination_port,
+                        chunks: vec![Box::new(err_chunk)]
+                    };
+                    vec![packet]
+                }
+                _ => unreachable!("This can only have 4 values.")
```

does this mean if someone sends an invalid handle code (evil peer?) we'll panic? Shouldn't we return an error instead?

```suggestion
                _ => return Err(Error::ErrChunkTypeUnhandled)
```

---

I get that the panic there might be bad, but it is also a bug in the libary if this case is hit. How do you generally handle these cases where a codepath should never be hit outside of a bug in the lib?

---

If the only case when we can get the unrecognised value here is bug in the library, then it's fine. I was thinking of malicious peers sending us invalid data for the purpose of DoS us.

---

`let handle_code = chunk.header().typ.0 >> 6;`
typ.0 is a u8, shifting it by 6 leaves us with those 4 bit patterns.

I'll leave it as it is then :)